src/index/addrindex.cpp

```diff
@@ -0,0 +1,263 @@
+
+  DBKeyPrefix() {}
+  explicit DBKeyPrefix(AddrId addr_id) :  m_addr_id(addr_id) {}
+
+  ADD_SERIALIZE_METHODS;
+
+  template <typename Stream, typename Operation>
+  inline void SerializationOp(Stream& s, Operation ser_action) {
+    char prefix = DB_ADDR_INDEX;
+    READWRITE(prefix);
+    if (prefix != DB_ADDR_INDEX) {
+      throw std::ios_base::failure("Invalid format for address index DB key");
+    }
+
+    READWRITE(m_addr_id);
+  }
+};
+
+struct DBKey : DBKeyPrefix {
+  DBKeyType m_key_type;
+  COutPoint m_outpoint;
+
+  DBKey() {}
+  explicit DBKey(DBKeyType key_type, AddrId addr_id, COutPoint outpoint) : DBKeyPrefix(addr_id), m_key_type(key_type), m_outpoint(outpoint) {}
+
+  ADD_SERIALIZE_METHODS;
+
+  template <typename Stream, typename Operation>
+  inline void SerializationOp(Stream& s, Operation ser_action) {
+    READWRITEAS(DBKeyPrefix, *this);
+
+    uint8_t key_type = static_cast<uint8_t>(m_key_type);
+    READWRITE(key_type);
+    m_key_type = static_cast<DBKeyType>(key_type);
+
+    // Check if the key type is a valid key. SEED keys are excluded because they
+    // are never created with this type.
+    if ((m_key_type != DBKeyType::SPENT) && (m_key_type != DBKeyType::CREATED)) {
+      throw std::ios_base::failure("Invalid key type for address index DB key");
+    }
+
+    READWRITE(m_outpoint);
+  }
+};
+
+}; // namespace
+
+// The address index stores information needed to get relevant transactions,
+// and a copy of the CScript to double check against in case of hash collisions.
+using DBValue = std::pair<CDiskTxPos, CScript>;
+
+/** Access to the addr_index database (indexes/addr_index/)*/
+class AddrIndex::DB : public BaseIndex::DB
+{
+public:
+    explicit DB(size_t n_cache_size, bool f_memory = false, bool f_wipe = false);
+
+    /** ReadAddrIndex returns the set of entries stored in the index for this addr_id. */
+    std::vector<std::pair<DBKey, DBValue>> ReadAddrIndex(const unsigned int addr_id, const CScript& script);
+
+    /** WriteToIndex writes the input vector of database entries into the index.  */
+    bool WriteToIndex(const std::vector<std::pair<DBKey, DBValue>> &entries);
+
+    /** SetupHashSeed is used to create/backup/restore the seed used by the index for hashing. */
+    unsigned int SetupHashSeed();
+};
+
+AddrIndex::DB::DB(size_t n_cache_size, bool f_memory, bool f_wipe) :
+    BaseIndex::DB(GetDataDir() / "indexes" / "addr_index", n_cache_size, f_memory, f_wipe)
+{}
+
+BaseIndex::DB& AddrIndex::GetDB() const { return *m_db; }
+
+std::vector<std::pair<DBKey, DBValue>> AddrIndex::DB::ReadAddrIndex(const unsigned int addr_id, const CScript& script)
+{
+    std::vector<std::pair<DBKey, DBValue>> result;
+    DBKeyPrefix search_key(addr_id);
+
+    std::unique_ptr<CDBIterator> iter(NewIterator());
+    iter->Seek(search_key);
+    while (iter->Valid()) {
+        DBKey key;
+        DBValue value;
+        if (!iter->GetKey(key) || key.m_addr_id != addr_id || !iter->GetValue(value) ) break;
```

Comment:
Do you want to log or throw something here? If I understand the LevelDB docs correctly, you have what LevelDB says is a valid iterator which should be positioned at the key. I think if any of these return false you'd want to know as the DB could be corrupted.

Reply:
I think in general it would be nice if this class did some of the error logging that I see in `blockfilterindex.cpp`.